### DISCUSSION TITLE
How can I mock the find_one function in each test?

### DISCUSSION THREAD

**Discussion by alexBonadei** [rond-authz/rond#163] [Q&A ~ The Runestone] [answered]
Hi,

I'm writing a policy and I'd like to test different cases that depend on the `find_one` function result.

Here's an example of the policy:
```
allow {
    userId := input.request.headers.userId

    user := find_one("users", { "userId": userId })

    user.userGroup == "admin"
}
```

This policy should pass when the retrieved user fromf `find_one` function is an admin and fail otherwise.

Therefore I'd like to test both cases.

Following the OPA documentation https://www.openpolicyagent.org/docs/latest/policy-testing/#data-and-function-mocking I wrote a test like this:

```
test {
 allow
    with input as {
      "request":{
        "headers":{
            "userid": "userId1"
            }
        }
    }
    with find_one as {    
    "userId": "userId1",
    "userGroup":"admin"
    }
}
```

The second `with` statement should mock the `find_one` function. But I get this error when running tests:
```
opa compiler fails to run tests: 1 error occurred: Test:34: rego_type_error: with keyword target must reference existing input, data, or a function
```

Am I missing something?
Thank you in advance

**Comment by fredmaggiowski** [comment, marked as answer]
Hi @alexBonadei 

as of now built-in mocking using the `with` keyword is only available for built-in defined by OPA; for "user-defined" built-ins (therefore the ones defined by rônd) you have to define the `find_one` function in your test file and eventually define different OR branches to supply different data on different tests.

```
package policies

find_one(collection, query) ret {
   collection == "users"
   ret := {some mock data for users}
} {
   collection == "books
   ret := {some mock data for books}
}

test_my_policy_using_find {
    my_policy_using_find
}
```

You can also make assertion on queries and differentiate the output of the mock based on the provided query.

**Reply by alexBonadei** [reply]
Ah ok I thought that it was possible to mock any function.

This way to defining the `find_one` mock should work for my purpose.

Thank you!
